Sources/Plasma/Apps/plClient/Mac-Cocoa/PLSPatcher.mm

```diff
@@ -139,13 +193,134 @@ bool IApproveDownload(const plFileName& file)
     return extExcludeList.find(file.GetFileExt()) == extExcludeList.end();
 }
 
+static la_ssize_t copy_data(struct archive *ar, struct archive *aw)
+{
+    la_ssize_t r;
+    const void *buff;
+    size_t size;
+    la_int64_t offset;
+    
+    while (true) {
+        r = archive_read_data_block(ar, &buff, &size, &offset);
+        if (r == ARCHIVE_EOF)
+            return (ARCHIVE_OK);
+        if (r < ARCHIVE_OK)
+            return (r);
+        r = archive_write_data_block(aw, buff, size, offset);
+        if (r < ARCHIVE_OK) {
+            fprintf(stderr, "%s\n", archive_error_string(aw));
+            return (r);
+        }
+    }
+}
+
+void Patcher::ISelfPatch(const plFileName& file)
+{
+    /*
+     Note on errors:
+     This function does not return errors, but a self patch
+     without a populated updatedClientURL will imply something
+     went wrong during decompress.
+     */
+    
+    PLSPatcher* patcher = parent;
+    patcher.selfPatched = true;
+    
+    int flags;
+    la_ssize_t r;
+
+    /* Select which attributes we want to restore. */
+    flags = ARCHIVE_EXTRACT_TIME;
+    flags |= ARCHIVE_EXTRACT_PERM;
+    flags |= ARCHIVE_EXTRACT_ACL;
+    flags |= ARCHIVE_EXTRACT_FFLAGS;
```

Comment:
Unclear if we need things like uid. The Mac installer model does maintain things like uid as part of the install process. It would be nice if we kept a consistent uid as the app was updated.

I'll take a look at the flags.

Reply:
Looking at the flags we're setting - if I take the Apple installer model of "all flags in a metadata tar should be honored" - I think it's possible that more flags may actually need to be added.

Part of this is I'm unsure what flags we'll need in the future - and both of Apple's installation formats (dmg and pkg) preserve flags. So I wouldn't want to disadvantage our patching against Apple's blessed technologies.

Reply:
Preserving the UID/GID from the archive would be actively bad, because we can't assume anything about the user running the game. No matter which UID/GID you put in the archive, some users will have a different one, so the only sensible choice is to extract with the UID/GID of the user running the game.

In any case, that behavior has nothing to do with these flags. The `ARCHIVE_EXTRACT_PERM` flag enables setting the [*set*uid, *set*gid](https://en.wikipedia.org/wiki/Setuid), and [sticky](https://en.wikipedia.org/wiki/Sticky_bit) permission bits, which are only relevant for certain system files/directories.

Generally, I'm cautious with preserving extra file metadata. I'm not aware of any recent case where applications are *required* to use specific file metadata, other than the execute bit. In fact, Apple [*disallows* certain file metadata in code-signed applications](https://developer.apple.com/library/archive/qa/qa1940/_index.html). Some of this metadata (especially xattrs) can appear unintentionally "out of nowhere", e. g. because you opened a folder in the Finder or copied/saved a file the wrong way, and that kind of metadata shouldn't be distributed to users.

Of course, it's impossible to predict what Apple will require in the future, but based on their current restrictions, I think it's safer to preserve less metadata and not more.

Reply:
> No matter which UID/GID you put in the archive, some users will have a different one, so the only sensible choice is to extract with the UID/GID of the user running the game.

This is built into how macOS handles installation. On every macOS system - the applications folder is always owned but the same UID. It's a constant across all macOS systems.

It is actually considered improper to ever use a user UID in the Applications folder.

Reply:
Unfortunately Apple has taken down most their documentation on PackageMaker and macOS installation, but I did find this:
https://developer.apple.com/library/archive/documentation/DeveloperTools/Conceptual/PackageMakerUserGuide/Workflow/Workflow.html#//apple_ref/doc/uid/TP40005371-CH1000-SW36

> Ownership and access permissions: The files that the Installer application places on the target computer have the same ownership and access permissions as the payload’s files. Therefore, you must set up the owner and access permissions of component files appropriately before building the installation package; otherwise, users may have difficulty manipulating those files after they are installed or Installer may be unable to copy payloads to their destinations

> In most cases, the owner should be root and the group admin. Also, PackageMaker can set the owner, group, and access permissions of the component files to those that work best in OS X, according to the component type.

One reason you want to install with a root owner is because the account that did the installation may disappear. In the case of proper macOS installers - the permissions will also be remembered by the Repair Permissions tool.

We'll need an installer for the initial installation - and that installer will install the application as owned by root - not the current user (which matches with Apple's guidance above.)

We could hardcode the patcher to just setuid everything to the root user. But that limits our flexibility just in case we need to assign another user. It would probably make sense to have the tar maintain the canonical permissions that were encoded in the tar - and handle this all on the packaging end.

Reply:
I've said this before, but I don't agree that Uru needs root ownership and a .pkg installer. In my experience, the most common installation method (outside the Mac App Store) is to distribute the application in a .dmg or .zip file and let the user copy it to the desired location, which is *usually* the system-wide Applications folder, but can also be any other folder. This means that the application will be owned by the user who installed it, but that's generally a good thing, because it reduces permission problems when updating/uninstalling later.

Yes, .pkg installers exist, but they're more complex and less flexible in many ways compared to .dmg/.zip distribution. In my experience, they're only used when a simple "drag to Applications" installation isn't enough, usually because the application needs something installed at the system level, like a driver, framework, or startup item. Uru doesn't need any of that, so we can simply distribute a .dmg/.zip instead.

> We could hardcode the patcher to just setuid everything to the root user.

I think we're still talking past each other here. I hope you're not suggesting that we should make the patcher a [setuid](https://en.wikipedia.org/wiki/Setuid#When_set_on_an_executable_file) executable - that would be a *massive* security risk.

Reply:
> I've said this before, but I don't agree that Uru needs root ownership and a .pkg installer. In my experience, the most common installation method (outside the Mac App Store) is to distribute the application in a .dmg or .zip file and let the user copy it to the desired location, which is usually the system-wide Applications folder, but can also be any other folder. This means that the application will be owned by the user who installed it, but that's generally a good thing, because it reduces permission problems when updating/uninstalling later.

I would prefer to not have a pkg installer as well. However there is an issue I have not been able to overcome without one.

.zip file distribution only works if your application is entirely self contained within it's bundle. Uru is not - its data files are stored outside the bundle. Because Uru is subject to [Gatekeeper Path Randomization](https://mjtsai.com/blog/2016/06/16/gatekeeper-path-randomization/) that means Uru is quarantined from accessing our own data or ini files - and it will fail to launch. The blog posts I've linked to also note this may break self patching - although I have no gotten that far because having our ini not work is pretty fatal.

There are only several ways around without using an installer AFAIK:
- Ensuring everyone uses a dmg for distribution. I don't know if this is better or worse than an installer. It leaves a lot of possible user error like trying to run from the dmg or not copying the entire Uru folder from the dmg.
- Distribute using a zip - but ensure the user moves the entire Uru application folder to another location to clear the quarantine. Again - could be error prone.

So far - an installer seems like (regrettably) the best path and matches the Windows behavior. If there was a way to have everything within our bundle and still support self patching that would be even more ideal - but does not seem possible.

> I think we're still talking past each other here. I hope you're not suggesting that we should make the patcher a [setuid](https://en.wikipedia.org/wiki/Setuid#When_set_on_an_executable_file) executable - that would be a massive security risk.

Ah - no. That is assuredly not what I meant.

Reply:
I was doing some more research on App Translocation tonight - and while I think it restates everything already linked - Quinn The Eskimo's write up is a good summary, and is also valuable for being from an Apple employees perspective.

https://forums.developer.apple.com/forums/thread/724969

Reply:
> Ensuring everyone uses a dmg for distribution.

Replying to myself to clarify - according to the documentation this needs to be a _signed_ dmg. Although I don't know if that means the flag will be cleared once the app is copied off of an unsigned dmg. We wouldn't support running from a dmg anyway.

Reply:
Oh - and the other core reason for an installer - we cannot let the user copy it into applications. If they do it will go into applications with the wrong permissions and it will not be able to update if an another user runs Uru.

While dmg allows the user to drag it anywhere they want - it would create improper behavior if the place they dragged it to was Applications. If we're going to be in a global location on disk the entire Uru folder needs to be world writable. (Again - we're not the only ones in this place. World of Warcraft has the same problem with the same solution.)

Reply:
Ah, I forgot about the app translocation thing - that's indeed going to be tricky. Though based on the things you've linked, distributing on a .dmg should be safe, because app translocation will be disabled when the user copies the game off of the disk image - right?

You're right that it's error-prone in the sense that the user might not copy all the files. I guess we could add one of those fancy background images to the .dmg, with a graphic explaining that you have to drag *all* the files off the image and not just the application.

The reason why I dislike the .pkg option so much is that .pkg installers are usually *really* inflexible. Most of them only let you install to a fixed system-wide location, and thus also require admin permissions for no good reason. Some allow you to install to the per-user Applications folder as an option (if you can spot the button for it), which lets you install without admin rights, but still doesn't allow multiple installations. It's *possible* to let the user change the installation path, but almost no installers bother supporting that, it seems...

> we cannot let the user copy it into applications. If they do it will go into applications with the wrong permissions and it will not be able to update if an another user runs Uru.

In my experience, when I copy something from a .dmg/.zip into Applications, it ends up with write permissions granted to my user and either the "admin" or "staff" group. So other admin users should still be able to update such apps without an elevation prompt, right? (Non-admin users can't, but as discussed before, that's the expected behavior for the system-wide Applications folder and I don't think we need/should work around that.)

Reply:
> (Non-admin users can't, but as discussed before, that's the expected behavior for the system-wide Applications folder and I don't think we need/should work around that.)

I don't think that's true - and I don't think there is agreement on that.

Uru would not be runnable at all by users without admin rights. That is not in line with expectations about Applications installed on a system or best practices. Applications installed in /Applications should be runnable by all users. Uru entangles updating and running - but that makes it our issue to deal with.

That's also how Uru on Windows deals with this issue. So by solving it the same way as Windows we'd have parity on Mac.

Reply:
> Uru would not be runnable at all by users without admin rights.

Even if Uru doesn't have write access to itself, it should still run as long as nothing needs patching, right? Not an optimal situation of course, because most shards update semi-regularly.

With the rest, I feel like we're going in circles :/ Honestly, it's also been a few years since I've actively used a Mac with multiple users, so I can't confirm if I'm understanding all the multi-user permission behaviors correctly.

But my main opinion is still: if installing Uru into /Applications causes permission problems for some users, and those problems can be avoided by installing Uru into a different folder, then we should just tell the affected users to do that. Anything else isn't worth the complexity IMO, unless the permission issues are really common.

Coming back to the original topic: regardless of how we deal with the permission issues, the patcher cannot enforce a specific owner UID/GID for the game files, because only root is allowed to `chown` files to other users. So even if we have an installer that installs all files as owned by root, that will be lost once the patcher updates the files.

And coming back to the *original* original topic: none of this changes that at least `ARCHIVE_EXTRACT_PERM` and `ARCHIVE_EXTRACT_ACL` should be removed from the flags, because they are potentially risky and have no legitimate use in our case.

Reply:
> Even if Uru doesn't have write access to itself, it should still run as long as nothing needs patching, right? Not an optimal situation of course, because most shards update semi-regularly.

I think there might need to be some clarification on how updates are deployed. As it stands - Uru patches itself while it's running. If you go to a new age that hasn't yet been downloaded - or link to an age that has updates - that patch may occur while the game is running and not part of the launch process. And that's what tends to cause problems.

If I only need to patch during the launch process - I could request admin privileges escalation from macOS at launch and perform the patch. The user would get the admin authorization dialog.

But because Uru can patch while it's running I'd have to put up an admin authorization dialog within age transition loading screens which is just not tenable. I don't think Uru has any handling for a privilege failure if it's downloading patch data as part of an age link.

AFAIK - in the wild shards don't force clients to patch all data on startup - but I could be wrong. From what I've seen the Uru launcher doesn't force all data to be patched at launch - only the essential data.

(World of Warcraft has a very similar system. It sometimes won't download a specific zone until you enter that zone. Which is why they also make their install directory world writable.)

> But my main opinion is still: if installing Uru into /Applications causes permission problems for some users, and those problems can be avoided by installing Uru into a different folder, then we should just tell the affected users to do that. Anything else isn't worth the complexity IMO, unless the permission issues are really common.

I think where things are at is if Uru is installed into /Applications with an installer - we can prevent problems for all users with all account types. Which I would prefer. I don't think we want to maintain an FAQ or have to walk users through post install steps for their specific configurations. I do know people who run under non-admin accounts for security reasons, and keep a logged out admin account they can use for privilege escalation when prompted.

I think the installer case doesn't preclude someone from installing to Applications, but then moving the Uru folder out of Applications. Mac installers support relocation. So I'd rather default to the safe behavior (install to /Applications) and if a user wants to take what was installed and move it elsewhere they are free to. We could even make a custom install directory part of the installer process.

I also have general dislike of installers on the Mac - but I think they do exist for a reason and apps like Uru might be that reason. All of Cyan's recent releases on Mac have also used installers so we'd be consistent with their other games.

> And coming back to the original original topic: none of this changes that at least ARCHIVE_EXTRACT_PERM and ARCHIVE_EXTRACT_ACL should be removed from the flags, because they are potentially risky and have no legitimate use in our case.

I've had a few things going on - not ignoring these changes, just haven't had a chance to look through yet.

Reply:
Yeah, I doubled check - at least for H'uru we default to on the fly patching. Which means patching occurs proactively while the user is playing and the client is running. The Image/kArgImage argument forces a full patch of all files - but it's not the default for new installs. Which means clients only get a core pack of files and not all of the ages.